### PR TITLE
Refactor getter typing logic and canonical signatures logic

### PR DESCRIPTION
**Pull request by cd1m0**
- Separate abi encoding logic in abi.ts;
- Move `getterArgsAndReturn` into `VariableDeclaration` class;
- Reduce code duplication between `getterArgsAndReturn` and `getterCanonicalSignature`;
- Add `signature.spec.ts` tests that compare the canonical types/signatures with the function hashes and ABI the compiler outputs
- Make ABIEncoder a required argument for all `canonicalSignature` entry points in all classes that have them. (WARNING: This is an API breaking change)

### CODE REVIEW

**Review by blitz-1306 — CHANGES_REQUESTED** [review]
Goot job. Just a few things to tweak / polish.

File: src/ast/definitions.ts [hunk 1]
```diff
@@ -213,12 +214,16 @@ function* lookupInContractDefinition(
                 let sigHash: string | undefined;
 
                 if (child instanceof FunctionDefinition) {
-                    sigHash = child.canonicalSignatureHash;
+                    // Its a safe to assume V2 as its abackward-compatible and
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/ast/definitions.ts]: Addressed in ed393c6.

File: src/ast/definitions.ts [hunk 2]
```diff
@@ -213,12 +214,16 @@ function* lookupInContractDefinition(
...
+                    // we only use it internally here
+                    sigHash = child.canonicalSignatureHash(ABIEncoderVersion.V2);
                 } else if (
                     child instanceof VariableDeclaration &&
                     child.visibility === StateVariableVisibility.Public
                 ) {
-                    sigHash = child.getterCanonicalSignatureHash;
+                    // Its a safe to assume V2 as its abackward-compatible and
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/ast/definitions.ts]: Addressed in ed393c6.

File: src/ast/implementation/declaration/variable_declaration.ts [hunk 3]
```diff
@@ -155,70 +170,104 @@ export class VariableDeclaration extends ASTNode {
...
+     * corresponding to this state variable.
+     */
+    getterArgsAndReturn(): [TypeNode[], TypeNode] {
+        const argTypes: TypeNode[] = [];
 
         let type = this.vType;
 
+        assert(
+            type !== undefined,
+            "Called getterArgsAndReturn() on variable thisaration without type",
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/ast/implementation/declaration/variable_declaration.ts]: Addressed in ed393c6.

File: src/bin/compile.ts [hunk 4]
```diff
@@ -267,6 +269,9 @@ OPTIONS:
 
     if (args.tree) {
         const INDENT = "|   ";
+
+        const encoderVersion = getABIEncoderVersion(units, result.compilerVersion as string);
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply src/bin/compile.ts]: Fixed it to not print signatures/interface ids when there is no known compiler version. Addressed in ed393c6.

File: test/samples/solidity/getters_08.sol [hunk 5]
```diff
@@ -45,4 +45,4 @@ contract AccessorReturns {
         function (uint256) external view returns (Some) g1 = this.g;
         function () external view returns (Some) h1 = this.h;
     }
-}
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply test/samples/solidity/getters_08.sol]: Addressed in ed393c6.

File: test/samples/solidity/getters_07_abiv1.sol [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        uint256[] memory arr = new uint[](3);
+        arr[0] = 1;
+        
+        uint x = this.a(0);
+        uint y = this.b(address(0x0));
+        E z  = this.c();
+        (E u, bytes1 v) = this.d();
+        address w = this.e();
+
+        Some t = this.f(0);
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply test/samples/solidity/getters_07_abiv1.sol]: Addressed in ed393c6.

File: test/samples/solidity/getters_07_abiv1.sol [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    }
+}
```
[the remark this answers is not in the document]
Comment (cd1m0) [reply test/samples/solidity/getters_07_abiv1.sol]: Addressed in ed393c6.

**Review by blitz-1306 — APPROVED** [review]
Looks good.

**Pull request merged**